harvester_e2e_tests/fixtures/virtualmachines.py

```diff
@@ -112,95 +136,253 @@ def __init__(self, vm_api, wait_timeout, snooze=3):
-                return False, (code, data)
+        @contextmanager
+        def configure(self, snooze=None, wait_timeout=None):
+            ''' context manager to temporarily change snooze or wait_timeout '''
+            s, t = self.snooze, self.wait_timeout
+            try:
+                self.snooze, self.wait_timeout = snooze or s, wait_timeout or t
+                yield self
+            finally:
+                self.snooze, self.wait_timeout = s, t
+
+        def wait_stopped(self, vm_name, endtime=None, callback=default_cb, **kws):
+            ctx = ResponseContext('vm.stop', *self.vms.stop(vm_name, **kws))
+            if 404 == ctx.code and callback(ctx):
+                return False, ctx
 
             endtime = endtime or self._endtime()
             while endtime > datetime.now():
-                code, data = self.vms.get_status(vm_name)
-                if 404 == code and callback(code, data):
+                ctx = ResponseContext('get_status', *self.vms.get_status(vm_name, **kws))
+                if 404 == ctx.code and callback(ctx):
                     break
                 sleep(self.snooze)
             else:
-                return False, (code, data)
-            return True, (code, data)
+                return False, ctx
+            return True, ctx
+
+        def wait_status_stopped(self, vm_name, endtime=None, callback=default_cb, **kws):
+            def cb(ctx):
+                if ctx.callee == 'vm.stop':
+                    return callback(ctx)
+                ctx.code, ctx.data = self.vms.get(vm_name, **kws)
+                ctx.callee = 'vm.get'
+                return (
+                    200 == ctx.code
+                    and "Stopped" == ctx.data.get('status', {}).get('printableStatus')
+                    and callback(ctx)
+                )
+            return self.wait_stopped(vm_name, endtime, cb, **kws)
+
+        def wait_deleted(self, vm_name, endtime=None, callback=default_cb, **kws):
+            ctx = ResponseContext('vm.delete', *self.vms.delete(vm_name, **kws))
+            if 404 == ctx.code and callback(ctx):
+                return False, ctx
+
+            endtime = endtime or self._endtime()
+            while endtime > datetime.now():
+                ctx = ResponseContext('vm.get_status', *self.vms.get_status(vm_name, **kws))
+                if 404 == ctx.code and callback(ctx):
+                    break
+                sleep(self.snooze)
+            else:
+                return False, ctx
+            return True, ctx
+
+        def wait_restarted(self, vm_name, endtime=None, callback=default_cb, **kws):
+            ctx = ResponseContext('vm.get_status', *self.vms.get_status(vm_name, **kws))
+            if 404 == ctx.code and callback(ctx):
+                return False, ctx
+
+            options = dict(old_pods=set(ctx.data['status']['activePods'].items()))
+            ctx = ResponseContext('vm.restart', *self.vms.restart(vm_name, **kws), options)
```

Comment:
`VMChecker` is designed to check whether the VM entered the state during the `wait_timeout`.
For example,  `checker.wait_stopped` will return **(vm_entered_state, context)** to indicate the result, It is not related to test logic but eliminate polling steps.  (that's why we won't do `assertion` inside)

As it is designed to check the state change, it should be fine to use other methods to get the expected state, it is also one of the reason to use `Context` object instead of return a tuple directly, the context will indicate the finally status and which API be called.

and that's why we are not checking status code in **204** but **404**, because the user (caller) will only care about the final state (VM stopped), and when we call `vm.stop`, it might return 204 or 500, but the VM will stopped after all.